go/ct/gen/block_context_test.go

```diff
@@ -54,56 +56,476 @@ func TestBlockContextGen_Generate(t *testing.T) {
+				} else if res.BlockNumber != 56 {
+					t.Errorf("expected block number to be 56, got %d", res.BlockNumber)
+				}
+			},
+		},
+		"variable-with-fixed-offset-beyond-uint64-range": {
+			setup: func(b *BlockContextGenerator, a Assignment) {
+				b.SetBlockNumberOffsetValue("a", -44)
+				a["a"] = common.NewU256(1, 12) // 2^64+12
+			},
+			check: func(t *testing.T, res st.BlockContext, a Assignment) {
+				if got, want := res.BlockNumber, common.NewU256(1, 12).Sub(common.NewU256(44)).Uint64(); got != want {
+					t.Errorf("expected block number to be %d, got %d", want, got)
+				}
+			},
+		},
+		"variable-in-range": {
+			setup: func(b *BlockContextGenerator, a Assignment) {
+				b.RestrictVariableToOneOfTheLast256Blocks("a")
+			},
+			check: func(t *testing.T, res st.BlockContext, a Assignment) {
+				value, present := a["a"]
+				if !present {
+					t.Errorf("expected variable 'a' to be assigned, got %v", a)
+				} else if !value.IsUint64() {
+					t.Errorf("value assigned to 'a' is out of range: %v", value)
+				} else {
+					value := value.Uint64()
+					if res.BlockNumber-value > 256 || res.BlockNumber-value < 1 {
+						t.Errorf("unexpected distance between variable 'a' and block number, got block number %d and assignment %d", res.BlockNumber, value)
+					}
+				}
+			},
+		},
+		"variable-in-range-with-pre-assigned-value": {
+			setup: func(b *BlockContextGenerator, a Assignment) {
+				b.SetRevision(common.R07_Istanbul)
+				b.RestrictVariableToOneOfTheLast256Blocks("a")
+				a["a"] = common.NewU256(800)
+			},
+			check: func(t *testing.T, res st.BlockContext, a Assignment) {
+				if !(800 < res.BlockNumber && res.BlockNumber < 1000) {
+					t.Errorf("expected block number to be in range 801-1000, got %d", res.BlockNumber)
+				}
+			},
+		},
+		"variable-out-of-range": {
+			setup: func(b *BlockContextGenerator, a Assignment) {
+				b.RestrictVariableToNoneOfTheLast256Blocks("a")
+			},
+			check: func(t *testing.T, res st.BlockContext, a Assignment) {
+				value, present := a["a"]
+				if !present {
+					t.Errorf("expected variable 'a' to be assigned, got %v", a)
+				} else if !value.IsUint64() {
+					// this is fine, the value is out of range
+				} else {
+					value := value.Uint64()
+					if !(res.BlockNumber-value > 256 || res.BlockNumber-value < 1) {
+						t.Errorf("unexpected distance between variable 'a' and block number, got block number %d and assignment %d", res.BlockNumber, value)
+					}
+				}
+			},
+		},
+		"variable-out-of-range-with-pre-assigned-value": {
+			setup: func(b *BlockContextGenerator, a Assignment) {
+				b.SetRevision(common.R07_Istanbul)
+				b.RestrictVariableToNoneOfTheLast256Blocks("a")
+				a["a"] = common.NewU256(8000)
+			},
+			check: func(t *testing.T, res st.BlockContext, a Assignment) {
+				if res.BlockNumber >= 1000 {
+					t.Errorf("produced block number is not a valid Istanbul block, got %d", res.BlockNumber)
+				}
+			},
+		},
```

Comment:
This test fails, though I think it should not:
```suggestion
		},
		"tight-block-range-with-single-solution": {
			setup: func(b *BlockContextGenerator, a Assignment) {
				b.blockNumberSolver = NewRangeSolver[uint64](0, 1)
				b.RestrictVariableToOneOfTheLast256Blocks("a")
			},
			check: func(t *testing.T, res st.BlockContext, a Assignment) {
				if res.BlockNumber != 1 {
					t.Errorf("expected block number to be 1, got %d", res.BlockNumber)
				}
				if value := a["a"]; value != common.NewU256(0) {
					t.Errorf("expected variable 'a' to be assigned to 0, got %v", value)
				}
			},
		},
```